jobs/extract_content/worker.go

```diff
@@ -0,0 +1,176 @@
+	})
+}
+
+type ExtractContentInterfaceImpl struct {
+	app *app.App
+}
+
+type ExtractContentWorker struct {
+	name        string
+	stopChan    chan struct{}
+	stoppedChan chan struct{}
+	jobsChan    chan model.Job
+	jobServer   *jobs.JobServer
+	app         *app.App
+	appContext  *request.Context
+}
+
+func (i *ExtractContentInterfaceImpl) MakeWorker() model.Worker {
+	return &ExtractContentWorker{
+		name:        "ExtractContent",
+		stopChan:    make(chan struct{}),
+		stoppedChan: make(chan struct{}),
+		jobsChan:    make(chan model.Job),
+		jobServer:   i.app.Srv().Jobs,
+		app:         i.app,
+		appContext:  &request.Context{},
+	}
+}
+
+func (w *ExtractContentWorker) JobChannel() chan<- model.Job {
+	return w.jobsChan
+}
+
+func (w *ExtractContentWorker) Run() {
+	mlog.Debug("Worker started", mlog.String("worker", w.name))
+
+	defer func() {
+		mlog.Debug("Worker finished", mlog.String("worker", w.name))
+		close(w.stoppedChan)
+	}()
+
+	for {
+		select {
+		case <-w.stopChan:
+			mlog.Debug("Worker received stop signal", mlog.String("worker", w.name))
+			return
+		case job := <-w.jobsChan:
+			mlog.Debug("Worker received a new candidate job.", mlog.String("worker", w.name))
+			w.doJob(&job)
+		}
+	}
+}
+
+func (w *ExtractContentWorker) Stop() {
+	mlog.Debug("Worker stopping", mlog.String("worker", w.name))
+	close(w.stopChan)
+	<-w.stoppedChan
+}
+
+func (w *ExtractContentWorker) doJob(job *model.Job) {
+	if claimed, err := w.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", w.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+
+	var err error
+	var fromTS int64 = 0
+	var toTS int64 = model.GetMillis()
+	if fromStr, ok := job.Data["from"]; ok {
+		if fromTS, err = strconv.ParseInt(fromStr, 10, 64); err != nil {
+			w.setJobError(job, model.NewAppError("ExtractContentWorker", "extrac_content.worker.do_job.invalid_input.from", nil, "", http.StatusBadRequest))
+			return
+		}
+		fromTS *= 1000
+	}
+	if toStr, ok := job.Data["to"]; ok {
+		if toTS, err = strconv.ParseInt(toStr, 10, 64); err != nil {
+			w.setJobError(job, model.NewAppError("ExtractContentWorker", "extrac_content.worker.do_job.invalid_input.to", nil, "", http.StatusBadRequest))
+			return
+		}
+		toTS *= 1000
+	}
+
+	var nFiles int
+	var nErrs int
+	for {
+		opts := model.GetFileInfosOptions{
+			Since:          fromTS,
+			SortBy:         model.FileinfoSortByCreated,
+			IncludeDeleted: false,
+		}
+		fileInfos, err := w.app.Srv().Store.FileInfo().GetWithOptions(0, 1000, &opts)
+		if err != nil {
+			w.setJobError(job, model.NewAppError("ExtractContentWorker", "extract_content.worker.do_job.file_info", nil, err.Error(), http.StatusInternalServerError))
+			return
+		}
+		if len(fileInfos) == 0 {
+			break
+		}
+		for _, fileInfo := range fileInfos {
+			if !ignoredFiles[fileInfo.Extension] {
```

Comment:
Is this best as a blacklist—as it currently is—or as a whitelist?

Reply:
Honestly I've just copied all that from the original place without giving it too much thought. Personally I'd probably prefer a list of allowed extensions to avoid troubles and unnecessary work but I can see how listing all supported document types may also become cumbersome. I'd like to hear @jespino as well.

Reply:
> Honestly I've just copied all that from the original place without giving it too much thought. Personally I'd probably prefer a list of allowed extensions to avoid troubles and unnecessary work but I can see how listing all supported document types may also become cumbersome. I'd like to hear @jespino as well.

Ah, sorry. It didn't occur to me that this wasn't new code. Sounds like this can all happen as a possible future improvement.